### DISCUSSION TITLE
What exactly is the new threading model in practice?

### DISCUSSION THREAD

**Discussion by ororsatti** [reactwg/react-native-new-architecture#197] [Q&A] [answered]
I have been experimenting with JSI and turbo modules lately. 
I thought this would be a magic solution for the performance issues but it seems that when I call a resources intensive function I still see a drop in the js thread.

Now,, that kinda makes sense, if we are no longer running things over a bridge, and accessing the cpp functions from the js thread, it makes sense if they are running in the js thread.

I'm just not sure, and the described treading model is pretty vague.

**Comment by cortinico** [comment, marked as answer]
What exactly is your question @ororsatti ?

With the New Architecture, you can access JSI which lets you practically "share memory" between JS and C++. That allows for performance critical applications to don't do message passing through the bridge, which ends up eating up a lot of performance.

> I thought this would be a magic solution for the performance issues

Nope, the New Architecture is not a "magic solution" in the sense that you just turn in on and have performance gains out of the box. You need to adapt your code to work with JSI.
A lot of popular libraries have already been adapted to use JSI and create experiences which were not possible without JSI/New Architecture (for example, https://github.com/mrousavy/react-native-vision-camera is one of them).

**Reply by ororsatti** [reply]
So my question is where a function that I defined in the cpp run, if it's exposed in the js ?

I knew my title was a bit off 

> What exactly is your question @ororsatti ?
> 
> With the New Architecture, you can access JSI which lets you practically "share memory" between JS and C++. That allows for performance critical applications to don't do message passing through the bridge, which ends up eating up a lot of performance.
> 
> > I thought this would be a magic solution for the performance issues
> 
> Nope, the New Architecture is not a "magic solution" in the sense that you just turn in on and have performance gains out of the box. You need to adapt your code to work with JSI.
> A lot of popular libraries have already been adapted to use JSI and create experiences which were not possible without JSI/New Architecture (for example, https://github.com/mrousavy/react-native-vision-camera is one of them).
